Source/data/iterators.hpp

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include "common.hpp"
+
+namespace devilution {
+class FieldsInRecordRange {
+	GetFieldResult *state_;
+	const char *const end_;
+
+public:
+	FieldsInRecordRange(GetFieldResult *state, const char *end)
+	    : state_(state)
+	    , end_(end)
+	{
+	}
+
+	class Iterator {
+		GetFieldResult *state_;
+		const char *const end_;
+
+	public:
+		using iterator_category = std::input_iterator_tag;
+		using value_type = std::string_view;
+
+		Iterator(const char *end)
+		    : state_(nullptr)
+		    , end_(end)
+		{
+		}
+
+		Iterator(GetFieldResult *state, const char *end)
+		    : state_(state)
+		    , end_(end)
+		{
+			*state_ = GetNextField(state_->next, end_);
+		}
+
+		bool operator==(const Iterator &rhs)
+		{
+			if (state_ == nullptr && rhs.state_ == nullptr)
+				return true;
+
+			return state_ != nullptr && rhs.state_ != nullptr && state_->next == rhs.state_->next;
+		}
+
+		bool operator!=(const Iterator &rhs)
+		{
+			return !(*this == rhs);
+		}
+
+		Iterator &operator++()
+		{
+			if (state_->endOfRecord()) {
+				state_ = nullptr;
+			} else {
+				*state_ = GetNextField(state_->next, end_);
+			}
+			return *this;
+		}
+
+		value_type operator*()
+		{
+			return state_->value;
+		}
+	};
+
+	Iterator begin() const
+	{
+		return { state_, end_ };
+	}
+
+	Iterator end() const
+	{
+		return { end_ };
+	}
+};
+
+class RecordsRange {
+	const char *const begin_;
+	const char *const end_;
+
+public:
+	RecordsRange(const std::string_view characters)
+	    : begin_(characters.data())
+	    , end_(characters.data() + characters.size())
+	{
+	}
+
+	class Iterator {
+		GetFieldResult state_;
+		const char *const end_;
+
+	public:
+		using iterator_category = std::forward_iterator_tag;
+		using value_type = FieldsInRecordRange;
+
+		Iterator(const char *end)

```

Comment:
Perhaps the constructor does not need arguments for the `end` iterator?

```c++
Iterator() : end_(nullptr) {}
```

Reply:
I've done exactly that but haven't pushed it yet 😅. Both iterators use GetFieldResult::Status::Uninitialised as a sentinel so that operator++ only scans if operator* hasn't been called at least once since the last advance. Reading a field as a number requires something like ostream::operator<< though and it's not a repeatable operation, needs a bit of care